source/db/dffml_source_db/source.py

```diff
@@ -0,0 +1,143 @@
+            await db_ctx.update(
+                self.parent.config.table_name,
+                key_value_pairs,
+                [[Condition("key", "=", record.key)]],
+            )
+            results = [
+                row
+                async for row in db_ctx.lookup(self.parent.config.table_name)
+            ]
+        self.logger.debug("update: %s", await self.record(record.key))
+
+    async def records(self) -> AsyncIterator[Record]:
+        async with self.parent.db_impl() as db_ctx:
+            async for result in db_ctx.lookup(self.parent.config.table_name):
+                yield self.convert_to_record(result)
+
+    def convert_to_record(self, result):
+        modified_record = {
+            "key": "",
+            "data": {"features": {}, "prediction": {}},
+        }
+        for key, value in result.items():
+            if key.startswith("feature_"):
+                modified_record["data"]["features"][
+                    key.replace("feature_", "")
+                ] = value
+            elif ("_value" in key) or ("_confidence" in key):
+                target = key.replace("_value", "").replace("_confidence", "")
+                modified_record["data"]["prediction"][target] = {
+                    "value": result[target + "_value"],
+                    "confidence": result[target + "_confidence"],
+                }
+            else:
+                modified_record[key] = value
+        return Record(modified_record["key"], data=modified_record["data"])
+
+    async def record(self, key: str):
+        record = Record(key)
+        async with self.parent.db_impl() as db_ctx:
+            try:
+                row = await db_ctx.lookup(
+                    self.parent.config.table_name,
+                    cols=None,  # None turns into *. We want all rows
+                    conditions=[[Condition("key", "=", key)]],
+                ).__anext__()
+            except StopAsyncIteration:
+                # This would happen if there is no matching row, so the async generator reached the end
+                return record
+
+        if row is not None:
+            features = {}
+            predictions = {}
+            for key, value in row.items():
+                if key.startswith("feature_"):
+                    features[key.replace("feature_", "")] = value
+                elif "_value" in key:
+                    target = key.replace("_value", "")
+                    predictions[target] = {
+                        "value": row[target + "_value"],
+                        "confidence": row[target + "_confidence"],
+                    }
+            record.merge(
+                Record(
+                    row["key"],
+                    data={"features": features, "prediction": predictions},
+                )
+            )
+        return record
+
+
+@entrypoint("dbsource")
+class DbSource(BaseSource):
+    CONFIG = DbSourceConfig
+    CONTEXT = DbSourceContext
+    db_impl: BaseDatabase
```

Comment:
This should be set at the instance level rather than the class level (so in `__init__`) . Also we could probably have it just be called `self.db`.